bin/network-monitor/src/faucet.rs

```diff
@@ -0,0 +1,240 @@
+///
+/// # Returns
+///
+/// The response from the faucet if successful, or an error if the test fails.
+async fn perform_faucet_test(
+    client: &Client,
+    faucet_url: &Url,
+) -> anyhow::Result<(GetTokensResponse, u64)> {
+    // Use a test account ID - convert to AccountId and format properly
+    let account_id = AccountId::try_from(ACCOUNT_ID_SENDER)
+        .context("Failed to create AccountId from test constant")?;
+
+    let account_id = account_id.to_string();
+    debug!("Generated account ID: {} (length: {})", account_id, account_id.len());
+
+    // Step 1: Request PoW challenge
+    let pow_url = faucet_url.join("/pow")?;
+    let response = client.get(pow_url).query(&[("account_id", &account_id)]).send().await?;
+
+    let response_text = response.text().await?;
+    debug!("Faucet PoW response: {}", response_text);
+
+    let challenge_response: PowChallengeResponse = serde_json::from_str(&response_text)
+        .with_context(|| format!("Failed to parse PoW response: {response_text}"))?;
+
+    debug!(
+        "Received PoW challenge: target={}, challenge={}...",
+        challenge_response.target,
+        &challenge_response.challenge[..16.min(challenge_response.challenge.len())]
+    );
+
+    // Step 2: Solve the PoW challenge
+    let nonce = solve_pow_challenge(&challenge_response.challenge, challenge_response.target)
+        .context("Failed to solve PoW challenge")?;
+
+    debug!("Solved PoW challenge with nonce: {}", nonce);
+
+    // Step 3: Request tokens with the solution
+    let tokens_url = faucet_url.join("/get_tokens")?;
+    let asset_amount = 1_000_000u64; // 1 token with 6 decimals
+
+    let response = client
+        .get(tokens_url)
+        .query(&[
+            ("account_id", account_id.as_str()),
+            ("is_private_note", "false"),
+            ("asset_amount", &asset_amount.to_string()),
+            ("challenge", &challenge_response.challenge),
+            ("nonce", &nonce.to_string()),
+        ])
+        .send()
+        .await?;
+
+    let response_text = response.text().await?;
+
+    let tokens_response: GetTokensResponse = serde_json::from_str(&response_text)
+        .with_context(|| format!("Failed to parse tokens response: {response_text}"))?;
+
+    Ok((tokens_response, challenge_response.target))
+}
+
+/// Solves a proof-of-work challenge using SHA-256 hashing.
+///
+/// # Arguments
+///
+/// * `challenge` - The challenge string in hexadecimal format.
+/// * `target` - The target value. A solution is valid if H(challenge, nonce) < target.
+///
+/// # Returns
+///
+/// The nonce that solves the challenge, or an error if no solution is found within reasonable
+/// bounds.
+fn solve_pow_challenge(challenge: &str, target: u64) -> anyhow::Result<u64> {
+    debug!("Solving PoW challenge: challenge={}, target={}", challenge, target);
+    // Try up to 10 million nonces - this should be reasonable for most difficulties
+    for nonce in 0..10_000_000u64 {
```

Comment:
10M attempts is ~23 bits of difficulty which should be solvable in about 10 seconds (though, depends a lot on the CPU). It is possible we go beyond that - I'd probably put this at 100M.